### DISCUSSION TITLE
[Question][Hw3] Getting <ty>'s from instructions

### DISCUSSION THREAD

**Discussion by Allgot** [prosyslab-classroom/cs424-program-reasoning#325] [Q&A]
Hello!
I need <ty> arguments from instructions, but I cannot find such functions..

Let's take a look at 'zext' instruction:
its syntax is defined as
```
<result> = trunc <ty> <value> to <ty2>
```
And I can get <value> using Llvm.operand function, but I can never get <ty>'s!
I've used oprands with 0, 1, 2 and failed; used param, also failed.

Is there any way to do so?

**Comment by pingpingy1** [comment]
I don't think the types are directly obtainable, but rather are already encoded in the `llvalue` object.
In `verifier.ml`, I think the distinction between integers and booleans is made with `Llvm.type_of var |> Llvm.integer_bitwidth`, which is 32 for integers and 1 for booleans.

**Comment by Allgot** [comment]
Thank you! I think I got an idea, thanks to you :)

But now I have an error, which says the condition of branch operation is regarded as an uninterpreted value.
```
Fatal error: exception Z3.Error("Uninterpreted '%cmp1' in <null>:\nif.then(#0,#1,#2,#3) :- \n entry(),\n (= (:var 3) %cmp),\n (= (:var 2) %call),\n (= (:var 1) %cmp1),\n (= (:var 0) %conv),\n (= %conv 0),\n (= %cmp (> %call 1)),\n (= %call %call),\n (= %cmp1 (> %call 10)),\n %cmp1,\n (= %cmp true).\n")
```

I think the cause must be related to the way of checking branch condition...
Specifically, a R1 in the readme file is defined as
```
R1: entry /\ (%cmp = %call > 1) /\ (%cmp1 = %call > 0) /\ %cmp1 => then(%call, %cmp, %cmp1)
```
so I simply used 'mk_and ~~~ Verifier.z3expr_of condition', which now I'm in doubt with. 

----

But actually there is another ambiguous thing, which is R3 in the same file:
```
R3: then(%call) => end(%call)
```
I thought the relation is parameterized by all of variables defined ahead, so this would be the right rule from my understanding.
```
R3: then(%call, %cmp, %cmp1) => end((%call, %cmp, %cmp1) 
```

----

Put together, I have two questions.
(1) Is it okay to simply put 'Verifier.z3expr_of condition'? That is, will the Z3 recognize this as the same variable from previous instructions?

(2) If so, why R3 of readme file is written only with %call variable? Why were %cmp and %cmp1 excluded from it?

Sorry for somewhat disordered questions, and thank you for reading!

**Comment by Hhro** [comment]
Hi,

(1) Yes, it will. `Verifier.z3expr_of` makes the Z3 expression for llvalue by calling the `Z3.Arithmetic.Integer.mk_const_s` or `Z3.Boolean.mk_const_s`, which makes constant expression with the name. 

```ocaml
let z3expr_of llvalue =
  match Llvm.classify_value llvalue with
  ...
  | _ -> (
      match Llvm.type_of llvalue |> Llvm.integer_bitwidth with
      | 32 ->
          let vname = Llvmutils.string_of_exp llvalue in
          Z3.Arithmetic.Integer.mk_const_s z3ctx vname
      | 1 -> Z3.Boolean.mk_const_s z3ctx (Llvmutils.string_of_exp llvalue)
      | _ -> failwith "invalid type")
```

For example, `Z3.Arithmetic.Integer.mk_const_s z3ctx "%call"` makes a expression of integer sort, whose name is "%call".

Since Z3 identifies the constant expressions by their name, you can expect that Z3 would recognize the variable.


(2) I think there was a mistake. The correct rule should be:
```
R0: entry

R1: entry /\ (%cmp = %call > 1) /\ (%conv = (ite %cmp 1 0)) /\ (%conv > 0) /\ (%cmp1 = %call > 0) 
/\ %cmp1 => then(%call, %cmp, %cmp1, %conv)

R2: entry /\ (%cmp = %call > 1) /\ (%conv = (ite %cmp 1 0)) /\ (%conv > 0) /\ (%cmp1 = %call > 0)
 /\ (not %cmp1) => else(%call, %cmp, %cmp1, %conv)

R3: then(%call, %cmp, %cmp1, %conv) => end(%call, %cmp, %cmp1, %conv)
```

**Comment by Allgot** [comment]
Thank you for detailed explanation!
But still I cannot undestand what the error means and why it happends...
So, for given example, my program added following rules(debug.smt2)
```
entry
(=> (and entry
         true
         (= %cmp (> %call 1))
         (= %conv (ite %cmp 1 0))
         (> %conv 0)
         (= %cmp1 (> %call 0))
         true
         true
         %cmp1)
    (if.then %conv %cmp1 %call %cmp))
(=> (and entry
         true
         (= %cmp (> %call 1))
         (= %conv (ite %cmp 1 0))
         (> %conv 0)
         (= %cmp1 (> %call 0))
         true
         true
         (not %cmp1))
    (if.else %conv %cmp1 %call %cmp))
(=> (and (if.then %conv %cmp1 %call %cmp) true true)
    (if.end %conv %cmp1 %call %cmp))
```
And it looks like being identical to what you supplied above, but Z3 raises an exception that %cmp1 is an uninterpreted variable in <null>:\nif.then(%conv, %cmp1, %call, %cml).
(The full error is same as above, which is
```
Fatal error: exception Z3.Error("
if.then(#0,#1,#2,#3) :- 
 entry(),
 (= (:var 3) %cmp),
 (= (:var 2) %call),
 (= (:var 1) %cmp1),
 (= (:var 0) %conv),
 (= %cmp (> %call 1)),
 (> %conv 0),
 (= %cmp1 (> %call 0)),
 %cmp1,
 (= %conv (ite %cmp 1 0)).
")
```
)

Could you explain why Z3 regards %cmp1 as uninterpreted variable, with given rules?

Thank you again!

**Comment by Hhro** [comment]
Your rule is almost right, but I think you forgot to add a quantifier to your rules. (use `Verifier.make_rule`)
```ocaml
let make_rule vars cons =
  if vars = [] then cons
  else
    let vars = ExprSet.of_list vars |> ExprSet.elements in
    Z3.Quantifier.mk_forall_const z3ctx vars cons None [] [] None None
    |> Z3.Quantifier.expr_of_quantifier
```
I guess Z3 thinks `%cmp` is uninterpreted since you don't give any sort to it. Using above function will help.

**Comment by Allgot** [comment]
Ohh... actually I didn't see why I should use 'make_rule' function but it play a crucial rule in our theorem proving task..!

Thank you so much for pointing it out :)
